Frontend

deerman31 opened the pull request:
lintやfmtなどが動くようになりました。
なので、フロントエンドの最低限の環境構築はできました。



これから、実装が始まります。

deerman31 committed:
next.jsをインストール
makeでlintとfmtを動くようにした

deerman31 committed:
フロントエンドのfmtやlint、typecheckはできるようにしました。

tachu2 commented on Makefile:
```diff
@@ -10,3 +13,39 @@ down:
+	cp -r ./srcs/app/frontend ./srcs/debug
+	@echo "FORMAT"
+	docker-compose -f ./srcs/debug_compose.yml run --name lint-container lint-format fmt
+	rm -rf ./srcs/app/frontend/src/
+	rm -rf ./srcs/app/backend/project/src/
```
これってrmしないといけないんでしたっけ？
cpされると勝手に上書きされるかとおもいました。

tachu2 commented on Makefile:
```diff
@@ -10,3 +13,39 @@ down:
+	rm -rf ./srcs/debug/backend
+
+
+# fmt:
+# 	@echo "FORMAT"
+# 	npm run format --prefix $(FRONTEND_DIR)
+# 	@echo ""
+# 	cargo fmt --manifest-path $(TOML_FILE)
+
+# lint:
+# 	@echo "LINT"
+# 	npm run lint --prefix $(FRONTEND_DIR)
+# 	@echo ""
+# 	@echo "TYPE-CHECK"
+# 	npm run type-check --prefix $(FRONTEND_DIR)
+# 	@echo ""
+# 	cargo clippy --manifest-path $(TOML_FILE)
```
不要なコードが残っています。

tachu2 reviewed (commented):
対応ありがとうございます！

deerman31 committed:
指摘された無駄なコメントアウトを削除
fmtの際の残ったcontainerの削除の処理を追加

make fmtでのdocker cp
docker cpだけでは、directoryの置き換えはできないため、rm -rf で削除してからcpしています。

tachu2 commented on srcs/lint-fmt/entrypoint.sh:
```diff
@@ -0,0 +1,29 @@
+
+set -e
+
+if [ "$1" = "fmt" ]; then
+    echo "Formatting frontend..."
+    #npm install --prefix $FRONTEND_DIR
+    npm run format --prefix $FRONTEND_DIR
+
+    echo "Formatting backend..."
+    cargo fmt --manifest-path $TOML_FILE
+
+
+elif [ "$1" = "lint" ]; then
+    echo "Linting frontend..."
+    npm run lint --prefix $FRONTEND_DIR
+
+    echo "Type-checking frontend..."
+    npm run type-check --prefix $FRONTEND_DIR
+
+
+    echo "Linting backend..."
+    cargo clippy --manifest-path $TOML_FILE
+
+else
+    echo "Unknown command: $1"
+    exit 1
+fi
```
イメージですが、以下の用に修正してあげるとみやすくていいと思います。
まだローカルで動かしていないですが多分動くと思います。
- main関数を作ってあげる
- エラーはエラー出力する
```sh
#!/bin/bash

set -e

err() {
  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $*" >&2
}

main() {
  local cmd="$1"
  if [[ ${cmd} != "fmt" && ${cmd} != "lint" ]]; then
    err "unknown command"
    return 1
  fi
  if [ "${cmd}" = "fmt" ]; then
    echo "Formatting frontend..."
    npm run format --prefix ${FRONTEND_DIR}

    echo "Formatting backend..."
    cargo fmt --manifest-path ${TOML_FILE}

  elif [ "${cmd}" = "lint" ]; then
    echo "Linting frontend..."
    npm run lint --prefix ${FRONTEND_DIR} || return 1

    echo "Type-checking frontend..."
    npm run type-check --prefix ${FRONTEND_DIR} || return 1

    echo "Linting backend..."
    cargo clippy --manifest-path ${TOML_FILE} || return 1
  fi
  return 0
}

main "$@"
```

deerman31 committed:
docker-composeを古いコマンドで実行していたため、修正

deerman31 committed:
entrypoint.shでteraoさんのアドバイスに従い、修正。

なお、backendとfrontendともにcompileのチェックする処理を追加しました。

deerman31 committed:
コメントアウトを削除

tachu2 reviewed (commented):
対応ありがとうございます。

next jsのbuild方法だけからないので、readmeを残すか、
buildを自動化（Makefileまたはscript）してほしいと思いました。
buildが自動化されるのが理想だと思います。

案としては、frontendのdirにMakefileなどがあってbuildができる

deerman31 committed:
まだ試せていないため、明日テストを予定していますが、一旦コミットします。

Makefile、replace_program.sh、replace.rsを追加しました。

Makefileで実行可能にしました。

anchor deployで生成されるProgram IDを必要なファイルに書き込むスクリプトとRustファイルを作成しました。

置き換えるファイルについては、replace_program_id.shに記述する必要があります。後ほど修正する可能性があります。

deerman31 committed:
Dockerfileでデバッグのためにexitしていたのを消し忘れていたので削除

deerman31 committed:
Dockerfileを再度修正

deerman31 committed:
コンパイルを通るように修正

Commit:
修正が必要。自身のｐｃで作業するためpush

deerman31 committed:
校舎pcでテスト

deerman31 committed:
grep.pyCOPYする処理をDockerfileで追加

Commit:
WORKDIRを変更し、cdせずにMakeできるようにしました。

tachu2 commented on Makefile:
```diff
@@ -10,3 +13,24 @@ down:
+	rm -rf ./srcs/lint-fmt/backend
+
+lint:
+	cp -r ./srcs/app/backend ./srcs/lint-fmt
+	cp -r ./srcs/app/frontend ./srcs/lint-fmt
+	@echo "LINT"
+	${DOCKER_COMPOSE} -f ./srcs/debug_compose.yml run --rm lint-format lint
+	rm -rf ./srcs/lint-fmt/frontend
+	rm -rf ./srcs/lint-fmt/backend
```
lintterはhost os側のファイルに変更を加えない（コーディングルールにそっているか確認するだけ）認識ですが、コピーしてrmする必要はありますか？

deerman31 replied:
34行目で行なっているcomposeの処理をする時にfrontendとbackendをコンテナにcpする必要があります。

しかし、lint-fmtディレクトリにはfrontend, backendがないため、
31, 32行目でappディレクトリからcpしています。

35,36行目でrmしているのは34行目でlintなどの処理が終わっており、必要ないためです。

fmtと見比べるとわかりますが、fmtの処理ではrmでも同様の処理をしています。

tachu2 commented on srcs/app/build_tool/grep.py:
```diff
@@ -0,0 +1,31 @@
+
+#def remove_str(line, remove_str) -> str:
+
+id = "Program Id: "
+path = "anchor_id.txt"
+
+s = find_str(path, id)
+if s == "":
+    print("anchor_id.txtに\"Program Id: \"がない")
+    exit(1)
+
+s = s.replace(id, '')
+
+with open(path, 'w') as file:
+    file.write(s)
```
スクリプトでもmain関数を作った方がみやすくていいと思います。

deerman31 replied:
次のpushで修正して、mainに書きました。

tachu2 commented on srcs/app/build_tool/grep.py:
```diff
@@ -0,0 +1,31 @@
+
+#def remove_str(line, remove_str) -> str:
+
+id = "Program Id: "
+path = "anchor_id.txt"
```
pythonで定数は、upper camelケースに統一するのがいいと思います。

deerman31 replied:
次のpushで修正しました。

deerman31 committed:
buildとlint, fmtを同じdocker-compose.ymlで行うようにしました。それに伴い、以下の点を変更しました。

1. appディレクトリにbuild_toolを作成し、そこにbuildに必要なDockerfileやスクリプトなどを置き、debug_toolにlint,fmtに必要なDockerfileやスクリプトを置いています。

2. 1の変更に伴い、各Dockerifleを修正しています。

deerman31 committed:
build_toolのreplace.rsをcargoで実行するようにした。

deerman31 committed:
appが原因ですか？

deerman31 committed:
buildの失敗、前々回のコミットの変更に伴う、Dockerfileの変更をしなかったから。

YungTatyu commented:
buildが失敗しました。
一旦報告まで。
```sh
=> ERROR [app 17/25] RUN npm run build                                                                                                                   14.9s
------                                                                                                                                                          
 > [app 17/25] RUN npm run build:                                                                                                                               
0.666                                                                                                                                                           
0.666 > my-nextjs-app@0.1.0 build                                                                                                                               
0.666 > next build                                                                                                                                              
0.666                                                                                                                                                           
1.504 Attention: Next.js now collects completely anonymous telemetry regarding usage.
1.504 This information is used to shape Next.js' roadmap and prioritize features.
1.504 You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
1.504 https://nextjs.org/telemetry
1.504 
1.591   ▲ Next.js 14.2.12
1.592 
1.669    Creating an optimized production build ...
12.12  ✓ Compiled successfully
12.12    Linting and checking validity of types ...
14.69 
14.69 Failed to compile.
14.69 
14.69 ./src/app/pages/api/checkEven.ts
14.69 13:1  Error: Delete `··`  prettier/prettier
14.69 16:1  Error: Delete `⏎`  prettier/prettier
14.69 
14.69 info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
------
failed to solve: process "/bin/sh -c npm run build" did not complete successfully: exit code: 1
make: *** [Makefile:10: up] Error 17
```

tachu2 commented on srcs/app/build_tool/grep.py:
```diff
@@ -0,0 +1,23 @@
+Id = "Program Id: "
+Path = "anchor_id.txt"
```
ごめんなさい。
pythonの定数はupper snake caseでした。
https://peps.python.org/pep-0008/#constants

YungTatyu commented:
buildしてコンテナが立ち上がり、next jsのページも見れました！
ありがとうございます。
buildがすべて自動化されていてすごくいいと思いました！

weijuan82113 reviewed (commented):
環境構築とファイルディレクトリの構造から理解したいと思います。
お主にbuild_toolとdebug_toolの役割を聞きたいです。

deerman31 committed:
pythonの定数をupper_camel_caseに変更

rikeda-cloud reviewed (approved):
mainのビルドとformatter, linterの３つを動かしてみました。
Next.jsのトップ画面も見れましたし、いい感じに動いていると思います。
LGTM!!
